### PR TITLE
TableView should cache created readers

### LINKED ISSUE #15166: Reader resource leakage in TableView
**Describe the bug**
When set `autoUpdatePartitionsInterval` to a small value like 1 second in `TableViewBuilder`, the table view client will keep create new reader (consumer) in each interval and causing resource leaks. 

The main reason is the TableViewImpl is not cache existing reader, see https://github.com/apache/pulsar/blob/04aa9e8e51869d1621a7e25402a656084eebfc09/pulsar-client/src/main/java/org/apache/pulsar/client/impl/TableViewImpl.java#L77-L81

**To Reproduce**
use pulsar 2.10.0, create TableView client with a small `autoUpdatePartitionsInterval`, and you will see the consumers are created in each loop from the logs.

### PR DESCRIPTION
**Pull request by nlu90**
Fixes #15166 

### Motivation

*Explain here the context, and why you're making that change. What is the problem you're trying to solve.*

### Modifications

Add the reader into `readers` concurrent after creation

### Verifying this change

- [ ] Make sure that the change passes the CI checks.

*(Please pick either of the following options)*

This change is a trivial rework / code cleanup without any test coverage.

*(or)*

This change is already covered by existing tests, such as *(please describe tests)*.

*(or)*

This change added tests and can be verified as follows:

*(example:)*
  - *Added integration tests for end-to-end deployment with large payloads (10MB)*
  - *Extended integration test for recovery after broker failure*

### Does this pull request potentially affect one of the following parts:

*If `yes` was chosen, please highlight the changes*

  - Dependencies (does it add or upgrade a dependency): (yes / no)
  - The public API: (yes / no)
  - The schema: (yes / no / don't know)
  - The default values of configurations: (yes / no)
  - The wire protocol: (yes / no)
  - The rest endpoints: (yes / no)
  - The admin cli options: (yes / no)
  - Anything that affects deployment: (yes / no / don't know)

### Documentation

Check the box below or label this PR directly.

Need to update docs? 

- [ ] `doc-required` 
(Your PR needs to update docs and you will update later)
  
- [X] `no-need-doc` 
Internal Implementation Details
  
- [ ] `doc` 
(Your PR contains doc changes)

- [ ] `doc-added`
(Docs have been already added)

### CODE REVIEW

**Review by Technoboy- — APPROVED** [review]

File: pulsar-client/src/main/java/org/apache/pulsar/client/impl/TableViewImpl.java [hunk 1]
```diff
@@ -224,9 +224,23 @@ private CompletableFuture<Reader<T>> newReader(String partition) {
                 .readCompacted(true)
                 .poolMessages(true)
                 .createAsync()
+                .thenCompose(this::cacheNewReader)
                 .thenCompose(this::readAllExistingMessages);
     }
 
+    private CompletableFuture<Reader<T>> cacheNewReader(Reader<T> reader) {
+        CompletableFuture<Reader<T>> future = new CompletableFuture<>();
+        if (this.readers.containsKey(reader.getTopic())) {
```
Comment (freeznet) [inline pulsar-client/src/main/java/org/apache/pulsar/client/impl/TableViewImpl.java:233]: do we need to close the reader if the readers already cached one?

**Review by Demogorgon314 — APPROVED** [review]

File: pulsar-client/src/main/java/org/apache/pulsar/client/impl/TableViewImpl.java (hunk 1 above)
Comment (gaozhangmin) [reply pulsar-client/src/main/java/org/apache/pulsar/client/impl/TableViewImpl.java:233]: Here, How about checking cache first before creating a new reader ?

**Review by gaozhangmin — CHANGES_REQUESTED** [review]

File: pulsar-client/src/main/java/org/apache/pulsar/client/impl/TableViewImpl.java (hunk 1 above)
Comment (gaozhangmin) [reply pulsar-client/src/main/java/org/apache/pulsar/client/impl/TableViewImpl.java:233]: Here, How about check  cache first before creating reader?

**Review by gaozhangmin — CHANGES_REQUESTED** [review]

**Review by Technoboy- — CHANGES_REQUESTED** [review]
Ah, this will cause the pre-created reader not to close.

**Comment by nlu90** [comment]
@freeznet @gaozhangmin @Technoboy- 

This `newReader()` method will only be called for newly discovered partitions, which implies there are no readers created for these partitions previously.

So the situation you mentioned won't happen in the code

**Review by Technoboy- — APPROVED** [review]
LGTM
@Jason918 @mattisonchao Could you please help review this patch ?

File: pulsar-client/src/main/java/org/apache/pulsar/client/impl/TableViewImpl.java (hunk 1 above)
Comment (gaozhangmin) [inline pulsar-client/src/main/java/org/apache/pulsar/client/impl/TableViewImpl.java:233]: Why  `readers.containsKey` checked twice?  the previous check is in `start` method
```
 // Add new Partitions
                    partitions.forEach(partition -> {
                        if (!readers.containsKey(partition)) {
                            futures.add(newReader(partition));
                        }
                    });
```

**Review by freeznet — APPROVED** [review]

File: pulsar-client/src/main/java/org/apache/pulsar/client/impl/TableViewImpl.java (hunk 1 above)
Comment (nlu90) [reply pulsar-client/src/main/java/org/apache/pulsar/client/impl/TableViewImpl.java:233]: This is safe-guard condition to make the future finish in both cases and for code completeness

**Review by gaozhangmin — APPROVED** [review]

**Pull request merged**